### PR TITLE
Enable BPF without disruption 

### PR DESCRIPTION
**Pull request by song-jiang**
## Description

This PR is based on https://github.com/tigera/operator/pull/2921. It should been seen as a continuation of the original work with the following modifications:

1. Fixed an issue for patching FelixConfigurations by using the correct variable. 
2. Rewrote most of the UTs.
3. Address @caseydavenport's comments from previous PR.

## For PR author

- [ ] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

File: pkg/controller/installation/bpf.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	match = match || annotationValue != nil && fc.Spec.BPFEnabled != nil && *annotationValue == *fc.Spec.BPFEnabled
+
+	if !match {
+		return errors.New(`Unable to set bpfEnabled: FelixConfiguration "default" has been modified by someone else, refusing to override potential user configuration.`)
+	}
+
+	return nil
+}
+
+// If the Installation resource has been patched to dataplane: BPF then the
```
Comment (caseydavenport) [inline pkg/controller/installation/bpf.go:56]: Just a nit, but the function comments should start with the function name. e.g.,

> isRolloutCompleteWithBPFVolumes checks if the calico-node DaemonSet . . .

File: pkg/controller/installation/core_controller.go [hunk 2]
```diff
@@ -1489,6 +1489,17 @@ func (r *ReconcileInstallation) Reconcile(ctx context.Context, request reconcile
 	// Tell the status manager that we're ready to monitor the resources we've told it about and receive statuses.
 	r.status.ReadyToMonitor()
 
+	// BPF Upgrade without disruption:
+	// Delegate logic implementation here using the state of the installation and dependent resources.
```
Comment (caseydavenport) [inline pkg/controller/installation/core_controller.go:1493]: I suggest we modify this comment to something like this:

> If eBPF is enabled in the operator API, patch FelixConfiguration to enable it within Felix.

**Review by caseydavenport — APPROVED** [review]
Two very minor comments, but otherwise this LGTM - thanks @song-jiang!

File: pkg/controller/installation/core_controller.go [hunk 3]
```diff
@@ -1696,7 +1706,75 @@ func (r *ReconcileInstallation) setDefaultsOnFelixConfiguration(install *operato
...
+				return false, err
+			} else {
+				updated = true
+			}
+		}
+	} else {
+		if fc.Spec.BPFEnabled == nil || *fc.Spec.BPFEnabled {
+			err := setBPFEnabledOnFelixConfiguration(fc, bpfEnabledOnInstall)
+			if err != nil {
+				reqLogger.Error(err, "Unable to enable eBPF data plane")
```
Comment (tomastigera) [inline pkg/controller/installation/core_controller.go:1769]: Is this correct? `bpfEnabledOnInstall == false` so arent's we trying to explicitly disable?

Comment (song-jiang) [reply pkg/controller/installation/core_controller.go:1769]: Good catch! Raised [PR](https://github.com/tigera/operator/pull/3307) to fix it. 
@caseydavenport PTAL. Thanks.

**Pull request merged**